### DISCUSSION TITLE
"array cannot contain ActiveRecord objects" -- Where is this stated?

### DISCUSSION THREAD

**Discussion by dchapman1988** [Shopify/maintenance_tasks#870] [Q&A]
👋 Hello all. I just recently wrote a task that iterates over an array, and to my surprised it errored out in production with the error "array cannot contain ActiveRecord objects". My question is, where in the documentation is this stated? I see in the README it states:

>  `collection`: return an Active Record Relation or an Array to be iterated over

But I'm not seeing a note that an Array must not contain any AR objects. Am I over looking this somewhere in the documentation?

**Comment by etiennebarrie** [comment]
It's a safeguard in `job-iteration`: https://github.com/Shopify/job-iteration/blob/v1.3.6/lib/job-iteration/enumerator_builder.rb#L56
It should error the same way in all environments though, nothing specific to production.

The idea is to catch when one loads Active Record objects and puts them in an Array, because once deployed to production that may lead to loading a few million records and that usually fails at some point.

Ideally you would express the array as a Relation directly, but as a workaround, if you know you're dealing with relatively few records, you can usually make a new relation, e.g.:

```ruby
def collection
  SomeRecord.where(id: array_of_records.map(&:id))
end
```

---

Feel free to contribute documentation around this. 🙏

**Reply by nvasilevski** [reply]
I just wanted to add that often if the task has an array of Active Record  objects (`array_of_records` in the example) -  these objects  were most likely built using a relation. So for example:

```ruby
def collection
  canada_users = User.where(country: "Canada").to_a
  User.where(id: canada_users.map(&:id))
end
```

Can be simplified to 
```ruby
def collection
  User.where(country: "Canada")
end
```

The only caveat is that in the second example job iteration will append `order by id limit 100` to build batches which sometimes may not be the most efficient query due to lack of indexes. But for most cases using an Active Record relation as a collection is the best practise.
